torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2111,10 +2111,31 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
 
   auto vh = v.conj().transpose(-2, -1);
 
+  // Based on a little experiment, min_threshold should be (approximately)
+  //   "max_abs_eigenvalue * (10 + matrix_size) * machine_eps".
+  // See: https://github.com/pytorch/pytorch/pull/50942
+  // Eps is still taking double and single eps value, because I don't know
+  //   if it would produce a lot of backward incompatibility due to large
+  //   eps for half-precision floating point
+  double eps = (lambda.dtype() == at::kDouble) ? std::numeric_limits<double>::epsilon()
+                                               : std::numeric_limits<float>::epsilon();
+  auto max_abs_lambda = lambda.abs().amax(/*dim=*/-1, /*keepdim=*/true).unsqueeze(-1);
+  auto min_threshold = max_abs_lambda * (10 + lambda.size(-1)) * eps;
+
   Tensor result;
   if (gv.defined()) {
       Tensor F = lambda.unsqueeze(-2) - lambda.unsqueeze(-1);
-      F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
+
+      // simple modification to make it numerically stable if there are
+      // repeated eigenvalues
+      // see https://github.com/pytorch/pytorch/issues/47599
+      // and https://arxiv.org/abs/2011.04366
+      // Using less-or-equal sign (instead of less than) so that when
+      // min_threshold is 0 (e.g. in half-precision), it behaves the same as
+      // the original algorithm (i.e. substituting only the diagonal with INF)
+      Tensor idx = F.abs() <= min_threshold;
+      F.index_put_({idx}, INFINITY);
```

Comment:
Could you please make tests with `max_{i, j} |lambda_j -  lambda_i| <= min_threshold + epsilon`.

Reply:
Again, could you please tell how you got this empirical result? Are you sure it does hold for well-conditioned matrices?

Reply:
I hope the new test in the latest commit answers these.